### ISSUE TITLE
Finetune possible? 

### ISSUE THREAD

**Issue by IlyaAb** [chuanyangjin/fast-DiT#2]
There is one global question: is it possible to train model on custom data, not ImageNet, with only one class, which is not in ImageNet?
So there are following questions:
1. Why do we need extract_features.py? This file doesn’t occur in original code from FB. Is it right that extract_features.py is extracting part from old train.py file for create features from images using vae encoder, and we dont need so much parameters in last strings in the end, like epochs, models? 
2. Is it possible to train model on custom data using ImageNet’s weights?

**Comment by SinanGncgl** [comment]
I am not the creator of one of the authors of the repos. But for your questions;
1. It is because of speeding up the process. You first extract so that later in the training part you do not have to deal with it.
2. Of course, just define your custom data loader and that's it. If it has only one class it means that you want to train it unconditionally, you can just get rid of the class parameter or just send the number of classes as 1. If you want to use Imagenet's weights just make sure that your data distribution is not that far from the imagenet so that fine-tuning works as expected, otherwise you will just fail the training.

**Comment by IlyaAb** [comment]
> I am not the creator or one of the authors of the repos. But for your questions;
> 
>     1. It is because of speeding up the process. You first extract so that later in the training part you do not have to deal with it.
> 
>     2. Of course, just define your custom data loader and that's it. If it has only one class it means that you want to train it unconditionally, you can just get rid of the class parameter or just send the number of classes as 1. If you want to use Imagenet's weights just make sure that your data distribution is not that far from the imagenet so that fine-tuning works as expected, otherwise you will just fail the training.

Thanks for the answer!
1. train.py, as I understand it, does not imply the ability to use the trained model, I added a couple of lines from sample.py to load weights for XL. In this case, the model will somehow adapt itself to one class, do I need to change model.y_embedder.embedding_table = torch.nn.Embedding(num_classes+1, 1152) by hand? If nothing is changed, then the collab gives an error with the size of the video memory, apparently with XL in the collab you can only do inference. There are no trained models of smaller sizes in the public domain right now, right? I also changed model.module.state_dict() to model.state_dict(), because 'module' at that position gives an error while training, I did not understand.
2. I’m trying to train models with sizes S and B from scratch, a dataset of faces with a size of 3k pictures, 50 epochs, as a result, when i make a samples is just colored noise, as if the model does not learn at all, although the code passes without errors... = torch.tensor([1000] * n, device=device), replaced 1000 with 1. Also changed class_labels=[207, 360, 387, 974, 88, 979, 417, 279] in sample.py to the class I have, and y_null = torch.tensor([1000] * n, device=device), replaced 1000 for 1. 
![sample_2023-07-24 12 17 19 071913](https://github.com/chuanyangjin/fast-DiT/assets/76262702/c1d25d04-70c9-45bc-8939-deea5bdebacf)

**Comment by IlyaAb** [comment]
Understand
1. For finetune add several strings for changing y_embedder in loaded weights, like: 
   t1 = torch.split(state_dict['y_embedder.embedding_table.weight'], 1)[895]   - choose class airplane for drone generation
   t2 = torch.split(state_dict['y_embedder.embedding_table.weight'], 1)[1000]
   state_dict['y_embedder.embedding_table.weight'] = torch.cat((t1, t2), 0)
Cannot finetune on free colab, needs more gpu memory, about 22-24 gb for ~1500 size dataset. 
2. Train from scratch S model needs more than 15k epoch for some results when 25% of images are of adequate quality. Train on one 3070ti f
![д6](https://github.com/chuanyangjin/fast-DiT/assets/76262702/9ea42f5c-9560-4e0b-81d1-7a39ce0acf9d)
or 3-4days summary, get loss 0.025 and this is not a plateau, training can be continued further
